File: Data Structures & Algorithms/invert-a-binary-tree/submission-2.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    public TreeNode InvertTree(TreeNode root) {
           //BFS
           //#############
           Queue<TreeNode> queue = new Queue<TreeNode>();
           if(root != null){
             queue.Enqueue(root);
           }

           while(queue.Count > 0){
               //process node -> swap
               TreeNode curr = queue.Dequeue();
               TreeNode temp = curr.left;
               curr.left = curr.right;
               curr.right = temp;
               
               //Add Children
               if(curr.left != null){
                  queue.Enqueue(curr.left);
               }
               if(curr.right != null){
                  queue.Enqueue(curr.right);
               }
           }

           return root;
    }
}

File: Data Structures & Algorithms/valid-binary-search-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    #N: we are given a binary tree
    #P: do bfs but add  tuple (node, lower, upper)
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        
        q = deque()
        if root:
           q.append( (root, float("-inf"), float("inf")) )
        
        while len(q):
            #process current
            node, low, high = q.popleft()
            if node.val <= low or node.val >= high:
               return False

            #add children
            if node.left:
               q.append( (node.left, low, node.val) )
            if node.right:
               q.append( (node.right, node.val, high) )   

        return True 
            
